### PR TITLE
Fix product shortcode pagination

### PR DESCRIPTION
**Pull request by claudiulodro**
Fixes the FINAL TODO mentioned in #17683

### CODE REVIEW

File: includes/wc-template-functions.php [hunk 1]
```diff
@@ -895,12 +893,10 @@ function woocommerce_result_count( $query = null ) {
...
 	 *
-	 * @param WP_Query $query Pass a query object or use global.
+	 * @param array $args Pass an associative array of parameters. Uses this if passed, otherwise uses global $wp_query.
 	 */
-	function woocommerce_catalog_ordering( $query = null ) {
-		if ( ! is_a( $query, 'WP_Query' ) ) {
-			$query = $GLOBALS['wp_query'];
-		}
+	function woocommerce_catalog_ordering( $args = array() ) {
+		$query = $GLOBALS['wp_query'];
```
Comment (mikejolley) [inline includes/wc-template-functions.php]: I'd switch the logic up here somewhat and in the other function so it's more readable.

- If $args is empty, inherit props from global query object.
- Include array of default args and use wp_parse_args
- Within the function, you can then use the variable names after parsing instead of checking isset multiple times. This also let's us whitelist what is used.

File: includes/wc-template-functions.php [hunk 2]
```diff
@@ -942,16 +938,14 @@ function woocommerce_catalog_ordering( $query = null ) {
...
 	 *
-	 * @param WP_Query $query Pass a query object or use global to pull pagination info.
+	 * @param array $args Pass an associative array of parameters. Uses this if passed, otherwise uses global $wp_query.
 	 */
-	function woocommerce_pagination( $query = null ) {
-		if ( ! is_a( $query, 'WP_Query' ) ) {
-			$query = $GLOBALS['wp_query'];
-		}
+	function woocommerce_pagination( $args = array() ) {
+		$query = $GLOBALS['wp_query'];
```
Comment (mikejolley) [inline includes/wc-template-functions.php]: See comment under woocommerce_catalog_ordering

File: includes/wc-template-functions.php (hunk 1 above)
Comment (mikejolley) [reply includes/wc-template-functions.php]: Also let's you change the variable naming if needed.

File: includes/shortcodes/class-wc-shortcode-products.php [hunk 3]
```diff
@@ -504,10 +505,18 @@ protected function get_products_ids() {
 			} else {
 				$products = new WP_Query( $this->query_args );
 			}
-			$ids = wp_parse_id_list( $products->posts );
+
+			$ids = array(
```
Comment (mikejolley) [inline includes/shortcodes/class-wc-shortcode-products.php]: Maybe name this $return now it's not just ids? Nicpick.

File: includes/shortcodes/class-wc-shortcode-products.php [hunk 4]
```diff
@@ -534,10 +543,10 @@ protected function product_loop() {
 
 		ob_start();
 
-		if ( $products_ids ) {
+		if ( $products_ids && $products_ids['ids'] ) {
```
Comment (mikejolley) [inline includes/shortcodes/class-wc-shortcode-products.php]: Since we are changing what `get_products_ids` returns, do we need a new method?

**Review by mikejolley — CHANGES_REQUESTED** [review]
Added feedback.

Did have an alternative idea though - just needs testing to see how much of a negative performance impact it would have...

If we stuck with queries (vs args), and cached IDs, could we pass the cached IDs into a new WP_Query object with `post__in`? If thats not slower than we have now we can get the best of both worlds.

File: includes/shortcodes/class-wc-shortcode-products.php (hunk 4 above)
Comment (claudiulodro) [reply includes/shortcodes/class-wc-shortcode-products.php]: I don't believe so. It's protected and it's only been around since 3.2.4. Before that it was called `get_products`. The only place it's used in in `product_loop` below.

Comment (claudiosanches) [reply includes/shortcodes/class-wc-shortcode-products.php]: I prefer we change the name of the method again.

File: includes/shortcodes/class-wc-shortcode-products.php (hunk 3 above)
Comment (claudiosanches) [reply includes/shortcodes/class-wc-shortcode-products.php]: Or just `$data` shoulds better.

**Comment by claudiulodro** [comment]
Good feedback. This should be ready to go.
- Refactored template functions to make them cleaner
- Renamed shortcode methods/variables to make them more correct
- Fixes phpcs stuff in wc-template-functions.php

**Review by claudiosanches — APPROVED** [review]
LGTM
